### ISSUE TITLE
Email arrival notifications - Enhancement request

### ISSUE THREAD

**Issue by CheechGe** [wavebox/waveboxapp#730]
* Wavebox Version: [e.g. 3.0.0]
3.14.7

* Operating System & Version: [e.g. Windows 10, Mac OSX 10.3.4, Ubuntu 17.10]
Windows 10

**Is your feature request related to a problem? Please describe.**
Yes, an operational problem not a technical one

Some information about what the problem is.
Email notifications are a nice feature however, there isn't an effective way to isolate
those that are more important from those that are not.

**Is your feature request related to an account type?**
For example, a Gmail account or Trello account.
GMail

**Describe the solution you'd like**
Information about what you want to happen or like to see.

This request would allow the user to apply customized filtering criteria to handle the notification process differently.
At minimum the enhancement would allow the filtering based on:
1. the source of the message (from address) and, 
2. the data contained in the Subject line.

The match parameters for the above should include:
1. contains
2. does not contain
3. is
4. isn't
5. begins with
6. ends with

An additional option to the above shall be to allow an expression builder (open source) to create more granular match criteria.

Notification actions.
Once the filtering mechanism has isolated a message the notification message the user shall be allowed to:
1. customize the message to be displayed
2. optionally specify an audio  file (.wav) be played.
3. specify:
	a. the amount of time the message is to be displayed or,
	b. the notification requires an actionable discharge (cancelled, closed, ...)
4. this notification process works independently from the standard WB notification options,
	in other words it will work even if the WB notifications are turned off.
	
Additional items:
1. if user supplied .wav files are used there will be a WB sub-folder where they must be placed.
2. there shall be a test process for the customizable notifications similar to the one provided
   but supports the additional features.

**Additional info**
Any other information or screenshots that might help us.
Currently I use Thunderbird as my email client and while I do not expect that WB have all of its features or those provided by TB add-ons because it is not an email system however, because WB does do the  email notifications therefore this request might apply.  Screenshots of the TB message filtering process and the Alert notifications provided by an add-on will be provided as an example in additional posts on this topic.f requested.

**Comment by Thomas101** [comment]
Hi, thanks for the detailed request. I think there are some things we can do here, especially for power users.

Firstly, have you tried the custom unread queries for Gmail & Google Inbox? I don't know if this would help, but it allows you to change what Wavebox considers unread and will then change the unread badge and notifications that get shown. [More info](https://wavebox.io/kb/custom-google-unread-counts). It's only for Gmail accounts though and I suspect may not do exactly what you're looking for - but worth a mention.

### Message matching
There are some things that are always standardized across the app (for example the account & account type that the notification originates from), however beyond this notifications vary considerably.

* Integrated accounts (Gmail, Ginbox, Outlook, Office 365, Trello) all have extended information such as the party who generated the notification, subject line etc so this would be easily searchable
* Slack - although integrated we use mixed mode system to bring up the messages. This means sometimes we have extended information and other times we just have the notification title and body
* Linked accounts (all others) in these we just have the notification title and message body

Depending on which notifications you want to match on and what you want to match we might have to have a plain text search or filtering.

### Notification actions
The support that each platform provides for notification varies considerably making some of these easier than others...

> 1. customize the message to be displayed

We can probably use some kind of templating to fill in the blanks on the messages

> 2. optionally specify an audio file (.wav) be played.

Platform support for this is patchy. Windows I believe you can. MacOS only built in sounds. Linux depends on built in libs. We could probably shim this though to play the sound separately from the os.

> 3. specify:
a. the amount of time the message is to be displayed or,
b. the notification requires an actionable discharge (cancelled, closed, ...)

Platform support patchy. Windows supports it, however previous experience with the node libraries to make use of it hasn't proved successful. MacOS supports. Linux doesn't.

> 4. this notification process works independently from the standard WB notification options,
in other words it will work even if the WB notifications are turned off.

This is all Wavebox side, so no problems.

---
 I think this is something we can investigate. Even if it's providing a partial feature set. Have you got a link for the thunderbird extenaion? It will help to see how it exposes these settings in a semi-sensible way!

Thanks

**Comment by CheechGe** [comment]
Hi Thomas:

Thank you for your response.

No I didn't try any settings on GMail as I've looked at them before I
didn't see anything similar to what I was asking for.  Besides as you noted
from what can determine this appears to be able to filter unread emails and
not specific emails.   Compared to Thunderbird Gmail filtering is very weak.

I have attached a SS of TB's filtering dialog box.  Under the first Drop
box there are various topics that can be selected (the default is
"Subject")..  While there are about 15 topic areas here I think only 2
would be necessary (for email notifications) which are Subject and From
which are available in the message header. The next drop down box allows
the user to pick the search criteria allowing the expressions; contains
(default), doesn't contain, is, isn't, begins with, ends with.  The third
is an Input window where the user types the search item.   Multiple filters
can be defined by clicking the "+" and defining other criteria.  Because
more than one filter criteria can be defined the user can select from; 1.
Match all of the following, 2. Match any of the following, and, 3 Match all
messages.

The section underneath the filtering parameters define the action to be
taken when there is a match (default Move Message to).  The Alert addon
adds a Mailbox Alert (MA) action to the standard TB list.  In other words
the Mailbox Alert ado-on uses the standard TB filtering mechanism to detect
the messages for which an alert is to be issued.

The second SS is the MA window which allows a user to define the alert to
be issue when there is a match.  The alert can be named (allowing for
multiple alerts for different messages) , a custom message can be defined
and an alert "sound" could be played.  Concerning the sound alerts, there
are several *(free) text-to-speech applications available where the user
can type a text message and have it converted to a .wav file.  The one I
use is Balabolka which is compatible with he MS speech platform, but as I
mentioned there are several others.

Lastly, to find the TB Alerts add do the search: Mailbox
Alert +thunderbird + addons.

Please let me know if I can be of further assistance.

F

**Comment by CheechGe** [comment]
Hi Thomas:

One other possibility if this were implemented.  The Mini-Menu has a Notifications section and while I would not like to see the Mailbox Alerts included in this section (IMO< mixing it with others defeats the purpose which is to isolate them from others) a separate section for these message would be useful and probably eliminate the need for a user activated dismissal process as part of the notification message.

Regards,

Frank

**Comment by Thomas101** [comment]
Hi, Github hasn't pulled the attachments out of the email reply. Is there any chance you can upload them? Thanks :)

**Comment by CheechGe** [comment]
Hi, as requested

Regards,

Frank


![tb_filter](https://user-images.githubusercontent.com/3653639/43140804-817efd02-8f22-11e8-9ab4-ff28a8c520c2.png)
![tb_alert](https://user-images.githubusercontent.com/3653639/43140819-8543a1fe-8f22-11e8-8daf-5101b8ba8b6c.png)
![tb_filter](https://user-images.githubusercontent.com/3653639/43140820-8559ac74-8f22-11e8-8fd3-44fe1d8c94ad.png)

**Comment by Thomas101** [comment]
Thanks, those are really helpful.

I'll have a word with the team and see what we can look at adding into Wavebox

**Comment by CheechGe** [comment]
You're welcome.

I appreciate your attention to it.

**Comment by CheechGe** [comment]
Hi Thomas

Just curious, I haven't seen any activity on this request,  is it still a possibility?  (not like you have nothing else to do  :>)  ).

**Comment by CheechGe** [comment]
Hi Thomas:

Just looking for a status on this request.  I'm using my other email system (Thunderbird) to handle these notifications that can not easily handled with Wavebox.

**Comment by CheechGe** [comment]
I haven't seen any mention on the status of this enhancement request.  Would just like to know if it is even possible/being considered?

Thx

**Comment by Thomas101** [comment]
Hi, this is something we're still looking at adding, but I don't have any specific timeframes for testing etc yet

**Comment by CheechGe** [comment]
Hi Thomas:

Thanks for the update.  Good to hear that it is at least on the ToDo list.

**Comment by CheechGe** [comment]
Hi Thomas:

Should I give up on the idea that this might happen?  It's really a PITA the have 2 email systems running.  Please let me know if not as I will need to come up with a different solution.

Thanks

Frank
